### PR TITLE
TST: Test aggregate with list values #25581

### LINKED ISSUE #25581: Inconsistent behavior when using GroupBy and pandas.Series.mode
#### Code Sample

```python
import pandas

# Works great
df1 = pandas.DataFrame([[20,'A'],[20,'B'],[10,'C']])
gb1 = df1.groupby(0).agg(pandas.Series.mode)
display(gb1)

# Exception: Must produce aggregated value
df2 = pandas.DataFrame([[20,'A'],[20,'B'],[30,'C']])
gb2 = df2.groupby(0).agg(pandas.Series.mode)
display(gb2)
```
#### Problem Description

As it seems, the above code works great for **df1**, returning the following result:
```
         1
0         
10       C
20  [A, B]
```
(where C is a *str*, and [A, B] is a *numpy.ndarray*)

However, it doesn't work for **df2**, throwing the following exception:
```
...
C:\ProgramData\Anaconda3\envs\py36\lib\site-packages\pandas\core\groupby\generic.py in _aggregate_named(self, func, *args, **kwargs)
    907             output = func(group, *args, **kwargs)
    908             if isinstance(output, (Series, Index, np.ndarray)):
--> 909                 raise Exception('Must produce aggregated value')
    910             result[name] = self._try_cast(output, group)
    911 

Exception: Must produce aggregated value
```

It looks like the order of the processing of the *agg function* affects pandas error checking; if the first result (GroupBy row) returns an *numpy.ndarray* - the above exception is thrown, but if the first result return a str/scalar - the processing continues and suppresses further such exceptions.

In my opinion, item may be related to #2656  or one of its root causes ("fast apply vs. old pathway" as they put it), and to #24016 (which fails to accept *numpy.ndarray* as a return value). However, in our case - where *pandas.Series.mode* sometimes returns a scalar and sometimes a *numpy.ndarray*, it is more illusive and more confusing and therefore inconsistent (I spent ~2 hours debugging this trying to understand why so many of my *agg function* calls work but only one doesn't).

#### Expected Output
```
         1
0         
20  [A, B]
30       C
```

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit: None
python: 3.6.4.final.0
python-bits: 64
OS: Windows
OS-release: 7
machine: AMD64
processor: Intel64 Family 6 Model 78 Stepping 3, GenuineIntel
byteorder: little
LC_ALL: None
LANG: None
LOCALE: None.None

pandas: 0.24.1
pytest: None
pip: 19.0.1
setuptools: 40.4.3
Cython: None
numpy: 1.15.2
scipy: 1.0.0
pyarrow: None
xarray: None
IPython: 6.2.1
sphinx: None
patsy: 0.5.0
dateutil: 2.6.1
pytz: 2017.3
blosc: None
bottleneck: None
tables: None
numexpr: None
feather: None
matplotlib: 3.0.0
openpyxl: 2.5.12
xlrd: 1.2.0
xlwt: None
xlsxwriter: None
lxml.etree: None
bs4: None
html5lib: 1.0.1
sqlalchemy: None
pymysql: None
psycopg2: None
jinja2: 2.10
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: None
gcsfs: None

</details>

### PR DESCRIPTION
**Pull request by srotondo**
- [ ] closes #25581
- [ ] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [ ] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

**Review by phofl — COMMENTED** [review]
you can run pre-commit locally to avoid these issues

File: pandas/tests/groupby/aggregate/test_aggregate.py [hunk 1]
```diff
@@ -1413,3 +1413,14 @@ def test_multi_axis_1_raises(func):
     gb = df.groupby("a", axis=1)
     with pytest.raises(NotImplementedError, match="axis other than 0 is not supported"):
         gb.agg(func)
+
+
+def test_agg_of_list():
+    # GH#25581
+    df1 = DataFrame([[20, "A"], [20, "B"], [30, "C"]])
```
Comment (phofl) [inline pandas/tests/groupby/aggregate/test_aggregate.py:1420]: Could you add the other cases from the issue too?

File: pandas/tests/groupby/aggregate/test_aggregate.py [hunk 2]
```diff
@@ -1413,3 +1413,14 @@ def test_multi_axis_1_raises(func):
     gb = df.groupby("a", axis=1)
     with pytest.raises(NotImplementedError, match="axis other than 0 is not supported"):
         gb.agg(func)
+
+
+def test_agg_of_list():
```
Comment (jreback) [inline pandas/tests/groupby/aggregate/test_aggregate.py:1418]: can you add mode in the test name as well

File: pandas/tests/groupby/aggregate/test_aggregate.py (hunk 1 above)
Comment (srotondo) [reply pandas/tests/groupby/aggregate/test_aggregate.py:1420]: From the issue there are 3 other test cases than the one I originally added.  I can add 2 of them, but the final test case still fails on the master with a different error message than in 1.4.1.

File: pandas/tests/groupby/aggregate/test_aggregate.py [hunk 3]
```diff
@@ -1413,3 +1413,25 @@ def test_multi_axis_1_raises(func):
...
+
+
+@pytest.mark.parametrize(
+    "test, constant",
+    [
+        ([[20, "A"], [20, "B"], [10, "C"]], {0: [10, 20], 1: ["C", ["A", "B"]]}),
+        ([[20, "A"], [20, "B"], [30, "C"]], {0: [20, 30], 1: [["A", "B"], "C"]}),
+        ([["a", 1], ["a", 1], ["b", 2], ["b", 3]], {0: ["a", "b"], 1: [1, [2, 3]]})
+        # This test case still fails
+        # ([["a", 1], ["a", 2], ["b", 3], ["b", 3]], {0: ["a", "b"], 1: [[1, 2], 3]})
```
Comment (mroeschke) [inline pandas/tests/groupby/aggregate/test_aggregate.py:1425]:
```suggestion
        pytest.param([["a", 1], ["a", 2], ["b", 3], ["b", 3]], {0: ["a", "b"], 1: [[1, 2], 3]}, marks=pytest.mark.xfail)
```

Comment (mroeschke) [reply pandas/tests/groupby/aggregate/test_aggregate.py:1425]: Then the comment above can also be removed

**Review by mroeschke — APPROVED** [review]

**Comment by mroeschke** [comment]
Thanks @srotondo

**Pull request merged**